File: src/spectrace/registry/schema.py
"""Registry YAML schema parsing for the R0 fixture family."""

from __future__ import annotations

from collections.abc import Callable, Mapping, Sequence
from typing import Any, TypeVar

from spectrace.registry.model import (
    Definition,
    EntityRegistry,
    ExpectedRange,
    ExpectedReferences,
    ExternalReference,
    RegistryDocument,
    RegistryEdge,
    RegistryEntity,
    RegistryLoadError,
)


Record = TypeVar("Record")


def parse_registry_data(raw_data: Any) -> EntityRegistry:
    """Convert parsed YAML data into trusted registry records."""

    data = _require_mapping(raw_data, "registry")
    entities = _parse_required_items(data.get("entities"), "entities", _parse_entity)
    _require_unique([entity.id for entity in entities], "entities[].id")
    _require_unique([entity.label for entity in entities], "entities[].label")

    return EntityRegistry(
        registry_version=_require_text(data.get("registryVersion"), "registryVersion"),
        document=_parse_document(data.get("document")),
        entities=entities,
        edges=_parse_required_items(data.get("edges"), "edges", _parse_edge),
        external_refs=_parse_optional_items(
            data.get("externalRefs"), "externalRefs", _parse_external_ref
        ),
    )


def _parse_document(raw_document: Any) -> RegistryDocument:
    document = _require_mapping(raw_document, "document")
    return RegistryDocument(
        id=_require_text(document.get("id"), "document.id"),
        title=_require_text(document.get("title"), "document.title"),
        path=_require_text(document.get("path"), "document.path"),
        fixture_family=_require_text(
            document.get("fixtureFamily"), "document.fixtureFamily"
        ),
        source_docs=_parse_required_text_items(
            document.get("sourceDocs"), "document.sourceDocs"
        ),
    )


def _parse_entity(raw_entity: Any, field: str) -> RegistryEntity:
    entity = _require_mapping(raw_entity, field)
    return RegistryEntity(
        id=_require_text(entity.get("id"), f"{field}.id"),
        label=_require_text(entity.get("label"), f"{field}.label"),
        type=_require_text(entity.get("type"), f"{field}.type"),
        defines=_parse_definition(entity.get("defines"), f"{field}.defines"),
        expected_references=_parse_expected_references(
            entity.get("expectedReferences"), f"{field}.expectedReferences"
        ),
    )


def _parse_definition(raw_definition: Any, field: str) -> Definition:
    definition = _require_mapping(raw_definition, field)
    return Definition(
        kind=_require_text(definition.get("kind"), f"{field}.kind"),
        text=_require_text(definition.get("text"), f"{field}.text"),
    )


def _parse_expected_references(raw_references: Any, field: str) -> ExpectedReferences:
    if raw_references is None:
        return ExpectedReferences()

    references = _require_mapping(raw_references, field)
    labels = _parse_optional_text_items(references.get("labels"), f"{field}.labels")
    ranges = _parse_optional_items(
        references.get("ranges"), f"{field}.ranges", _parse_expected_range
    )
    return ExpectedReferences(labels=labels, ranges=ranges)


def _parse_expected_range(raw_range: Any, field: str) -> ExpectedRange:
    expected_range = _require_mapping(raw_range, field)
    return ExpectedRange(
        label_family=_require_text(
            expected_range.get("labelFamily"), f"{field}.labelFamily"
        ),
        start=_require_text(expected_range.get("start"), f"{field}.start"),
        end=_require_text(expected_range.get("end"), f"{field}.end"),
        expands_to=_parse_required_text_items(
            expected_range.get("expandsTo"), f"{field}.expandsTo"
        ),
    )


def _parse_edge(raw_edge: Any, field: str) -> RegistryEdge:
    edge = _require_mapping(raw_edge, field)
    return RegistryEdge(
        source=_require_text(edge.get("from"), f"{field}.from"),
        relationship=_require_text(edge.get("relationship"), f"{field}.relationship"),
        target=_require_text(edge.get("to"), f"{field}.to"),
    )


def _parse_external_ref(raw_external_ref: Any, field: str) -> ExternalReference:
    external_ref = _require_mapping(raw_external_ref, field)
    return ExternalReference(
        system=_require_text(external_ref.get("system"), f"{field}.system"),
        key=_require_text(external_ref.get("key"), f"{field}.key"),
        related_entity=_require_text(
            external_ref.get("relatedEntity"), f"{field}.relatedEntity"
        ),
        role=_require_text(external_ref.get("role"), f"{field}.role"),
    )


def _require_mapping(value: Any, field: str) -> Mapping[str, Any]:
    if not isinstance(value, Mapping):
        raise RegistryLoadError(f"{field} must be a mapping")
    return value


def _require_sequence(value: Any, field: str) -> Sequence[Any]:
    if isinstance(value, str) or not isinstance(value, Sequence):
        raise RegistryLoadError(f"{field} must be a list")
    return value


def _require_text(value: Any, field: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise RegistryLoadError(f"{field} must be a non-empty string")
    return value


def _parse_required_items(
    value: Any,
    field: str,
    parser: Callable[[Any, str], Record],
) -> tuple[Record, ...]:
    return _parse_items(_require_sequence(value, field), field, parser)


def _parse_optional_items(
    value: Any,
    field: str,
    parser: Callable[[Any, str], Record],
) -> tuple[Record, ...]:
    if value is None:
        return ()
    return _parse_items(_require_sequence(value, field), field, parser)


def _parse_required_text_items(value: Any, field: str) -> tuple[str, ...]:
    return _parse_text_items(_require_sequence(value, field), field)


def _parse_optional_text_items(value: Any, field: str) -> tuple[str, ...]:
    if value is None:
        return ()
    return _parse_text_items(_require_sequence(value, field), field)


def _parse_items(
    values: Sequence[Any],
    field: str,
    parser: Callable[[Any, str], Record],
) -> tuple[Record, ...]:
    return tuple(
        parser(item, f"{field}[{index}]") for index, item in enumerate(values)
    )


def _parse_text_items(values: Sequence[Any], field: str) -> tuple[str, ...]:
    return tuple(
        _require_text(item, f"{field}[{index}]")
        for index, item in enumerate(values)
    )


def _require_unique(values: Sequence[str], field: str) -> None:
    seen: set[str] = set()
    for value in values:
        if value in seen:
            raise RegistryLoadError(f"{field} contains duplicate value {value!r}")
        seen.add(value)

File: src/markdowntrace/registry/schema.ts
import {
  EntityRegistry,
  RegistryLoadError,
  type Definition,
  type ExpectedRange,
  type ExpectedReferences,
  type ExternalReference,
  type RegistryDocument,
  type RegistryEdge,
  type RegistryEntity,
} from "./model.js";

type RawRecord = Record<string, unknown>;
type ItemParser<T> = (value: unknown, field: string) => T;

export function parseRegistryData(rawData: unknown): EntityRegistry {
  const data = requireRecord(rawData, "registry");
  const entities = parseRequiredItems(data.entities, "entities", parseEntity);

  requireUnique(
    entities.map((entity) => entity.id),
    "entities[].id",
  );
  requireUnique(
    entities.map((entity) => entity.label),
    "entities[].label",
  );

  return new EntityRegistry({
    registryVersion: requireText(data.registryVersion, "registryVersion"),
    document: parseDocument(data.document),
    entities,
    edges: parseRequiredItems(data.edges, "edges", parseEdge),
    externalRefs: parseOptionalItems(data.externalRefs, "externalRefs", parseExternalRef),
  });
}

function parseDocument(rawDocument: unknown): RegistryDocument {
  const document = requireRecord(rawDocument, "document");

  return {
    id: requireText(document.id, "document.id"),
    title: requireText(document.title, "document.title"),
    path: requireText(document.path, "document.path"),
    fixtureFamily: requireText(document.fixtureFamily, "document.fixtureFamily"),
    sourceDocs: parseRequiredTextItems(document.sourceDocs, "document.sourceDocs"),
  };
}

function parseEntity(rawEntity: unknown, field: string): RegistryEntity {
  const entity = requireRecord(rawEntity, field);

  return {
    id: requireText(entity.id, `${field}.id`),
    label: requireText(entity.label, `${field}.label`),
    type: requireText(entity.type, `${field}.type`),
    defines: parseDefinition(entity.defines, `${field}.defines`),
    expectedReferences: parseExpectedReferences(
      entity.expectedReferences,
      `${field}.expectedReferences`,
    ),
  };
}

function parseDefinition(rawDefinition: unknown, field: string): Definition {
  const definition = requireRecord(rawDefinition, field);

  return {
    kind: requireText(definition.kind, `${field}.kind`),
    text: requireText(definition.text, `${field}.text`),
  };
}

function parseExpectedReferences(rawReferences: unknown, field: string): ExpectedReferences {
  if (rawReferences === undefined || rawReferences === null) {
    return { labels: [], ranges: [] };
  }

  const references = requireRecord(rawReferences, field);

  return {
    labels: parseOptionalTextItems(references.labels, `${field}.labels`),
    ranges: parseOptionalItems(references.ranges, `${field}.ranges`, parseExpectedRange),
  };
}

function parseExpectedRange(rawRange: unknown, field: string): ExpectedRange {
  const expectedRange = requireRecord(rawRange, field);

  return {
    labelFamily: requireText(expectedRange.labelFamily, `${field}.labelFamily`),
    start: requireText(expectedRange.start, `${field}.start`),
    end: requireText(expectedRange.end, `${field}.end`),
    expandsTo: parseRequiredTextItems(expectedRange.expandsTo, `${field}.expandsTo`),
  };
}

function parseEdge(rawEdge: unknown, field: string): RegistryEdge {
  const edge = requireRecord(rawEdge, field);

  return {
    source: requireText(edge.from, `${field}.from`),
    relationship: requireText(edge.relationship, `${field}.relationship`),
    target: requireText(edge.to, `${field}.to`),
  };
}

function parseExternalRef(rawExternalRef: unknown, field: string): ExternalReference {
  const externalRef = requireRecord(rawExternalRef, field);

  return {
    system: requireText(externalRef.system, `${field}.system`),
    key: requireText(externalRef.key, `${field}.key`),
    relatedEntity: requireText(externalRef.relatedEntity, `${field}.relatedEntity`),
    role: requireText(externalRef.role, `${field}.role`),
  };
}

function requireRecord(value: unknown, field: string): RawRecord {
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    throw new RegistryLoadError(`${field} must be a mapping`);
  }

  return value as RawRecord;
}

function requireList(value: unknown, field: string): readonly unknown[] {
  if (!Array.isArray(value)) {
    throw new RegistryLoadError(`${field} must be a list`);
  }

  return value;
}

function requireText(value: unknown, field: string): string {
  if (typeof value !== "string" || value.trim() === "") {
    throw new RegistryLoadError(`${field} must be a non-empty string`);
  }

  return value;
}

function parseRequiredItems<T>(
  value: unknown,
  field: string,
  parser: ItemParser<T>,
): readonly T[] {
  return parseItems(requireList(value, field), field, parser);
}

function parseOptionalItems<T>(
  value: unknown,
  field: string,
  parser: ItemParser<T>,
): readonly T[] {
  if (value === undefined || value === null) {
    return [];
  }

  return parseItems(requireList(value, field), field, parser);
}

function parseRequiredTextItems(value: unknown, field: string): readonly string[] {
  return parseTextItems(requireList(value, field), field);
}

function parseOptionalTextItems(value: unknown, field: string): readonly string[] {
  if (value === undefined || value === null) {
    return [];
  }

  return parseTextItems(requireList(value, field), field);
}

function parseItems<T>(
  values: readonly unknown[],
  field: string,
  parser: ItemParser<T>,
): readonly T[] {
  return values.map((item, index) => parser(item, `${field}[${index}]`));
}

function parseTextItems(values: readonly unknown[], field: string): readonly string[] {
  return values.map((item, index) => requireText(item, `${field}[${index}]`));
}

function requireUnique(values: readonly string[], field: string): void {
  const seen = new Set<string>();

  for (const value of values) {
    if (seen.has(value)) {
      throw new RegistryLoadError(`${field} contains duplicate value '${value}'`);
    }

    seen.add(value);
  }
}
